articles/Frontend/Architecture/good-practices.md

```diff
@@ -0,0 +1,1724 @@
+                this.render(container);
+            }           
+        })
+        .on('render', function onRender() {
+            this.getElement().on('click', () => this.click());                 
+        })
+        .init(config);
+}
+
+// Only a single way of reacting to click actions
+button('.fixture', {label: 'Ok'})
+    .on('click', () => { ... });
+```
+
+#### Resources
+- [Coding guide: General rules](coding-guide.md#general-rules)
+
+## Testing
+
+### Prefer design by coding: TDD / BDD
+When writing code, a developer must check the works is going the right way.
+Usually, some manual checks are made aside. But it may become more and more
+time consuming as long as the progress is made. So it is better to automate
+somehow those manual checks. And on convenient way to do so is to write
+unit tests.
+
+A method well describes the process, it is called [Test Driven Development](https://en.wikipedia.org/wiki/Test-driven_development).
+The principle is to write tests aside at the same period of time you are
+implementing. The best being to start writing test, then implement the feature.
+
+It might seem tricky to start writing unit tests while you have no idea how
+you will implement the feature. However, some patterns might help you.
+
+For instance, if you intent to write a component, since this is a well know
+pattern you might start by bootstrapping a common test pattern applied to 
+components: check the format of the component factory, check the common API, 
+prepare the check of the life cycle, and add a visual test. Then you may 
+start implementing the component, adding more unit tests each time you 
+augment the component implementation.
+
+Another way of doing TDD might be to design by coding. If you know what you
+intent to implement, start by drafting the client, the code that will consume
+the feature. This way you will draft out the implementation. Take a look at the
+resource video shared in the resources section below.  
+
+#### Resources
+- [Design by Coding - YouTube video](https://www.youtube.com/watch?v=d5Y1B1cmaGQ)
+
+### Properly scope test fixtures
+Unit tests must be unique, predictable, and reproducible. They must not be
+dependant to other tests, and must not conflict as well. A common mistake is to
+rely on the same markup to render content for each unit test. This might be ok
+when tests are ran in series, one at a time. In that case the place must be
+cleaned between each iteration. However, when the tests are ran in parallel
+there is a big chance they will conflict, and will produce inconsistent results.
+
+Having inconsistent tests might be the symptom of concurrency issue within
+the test suite.
+
+To prevent any conflict, and to make sure each unit test is properly scoped
+this is a good habit to reserve one unique markup for each different test. 
+Keeping the same markup for a serie of tests is still legit however, as they 
+will run one after the other. The requirement being to clean the markup 
```

Comment:
You can actually reuse the same fixtures between tests. 
After each test the fixture element is removed and created again, so if you all your tests relies on the same structure you can safely reuse the fixture in the next test.

Reply:
Yes indeed, the fixture is cleaned by the test engine. However, it may also run the tests in parallel, and in that case the tests will conflict if they expect incompatible changes.

Reply:
as far as I know QUnit guarantee the DOM of the fixture will be the defined DOM for each and every test case (and doesn't run the tests in parallel, can be in a different order but it doesn't support parallel tests)

Reply:
Ok, my bad.
I rephrased the section